src/tools/miri/src/shims/unix/linux/sync.rs

```diff
@@ -145,13 +175,21 @@ pub fn futex<'tcx>(
             // It's not uncommon for `addr` to be passed as another type than `*mut i32`, such as `*const AtomicI32`.
             let futex_val = this.read_scalar_atomic(&addr, AtomicReadOrd::Relaxed)?.to_i32()?;
             if val == futex_val {
+                // Check that the top waiter (if exists) is waiting using FUTEX_WAIT_*.
```

Comment:
Please explain why we do this. Is there a test covering this case?

Reply:
According to the manpage:

```
       EINVAL (FUTEX_WAKE, FUTEX_WAKE_OP, FUTEX_WAKE_BITSET,
              FUTEX_REQUEUE, FUTEX_CMP_REQUEUE) The kernel detected an
              inconsistency between the user-space state at uaddr and
              the kernel state—that is, it detected a waiter which waits
              in FUTEX_LOCK_PI or FUTEX_LOCK_PI2 on uaddr.

       EINVAL (FUTEX_LOCK_PI, FUTEX_LOCK_PI2, FUTEX_TRYLOCK_PI,
              FUTEX_UNLOCK_PI) The kernel detected an inconsistency
              between the user-space state at uaddr and the kernel
              state.  This indicates either state corruption or that the
              kernel found a waiter on uaddr which is waiting via
              FUTEX_WAIT or FUTEX_WAIT_BITSET.
```

It results in EINVAL if these two series of op are mix-used. So I add detection for such scenario.